lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,657 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+
+  /// Retrieve execution count for a basic block
+  Optional<uint64_t> getExecutionCount(const BasicBlock* BB) const{
+    Function* Func = const_cast<Function*>(BB->getParent());
+    return LookupBFI(*Func)->getBlockProfileCount(BB);
+  }
+
+  /// Retrive a pair of information if the instruction is accessing any struct type and field number
+  Optional<std::pair<const Type*, unsigned> > getFieldAccessOnInstruction(const Instruction* I) const;
+
+  /// Print all accesses of all struct types defined in the program
+  void debugPrintAllStructAccesses();
+
+  /// Print the IR of the module with annotated information about struct access
+  void debugPrintAnnotatedModule();
+
+  /// Increment stats for one category
+  void addStats(unsigned Category) { StatCounts[Category]++; }
+  /// Print a brief stats of struct access
+  void printStats();
+
+ private:
+  const Module& CurrentModule;
+
+  /// Function reference that is used to retrive execution count for basic block
+  function_ref<BlockFrequencyInfo *(Function &)> LookupBFI;
+
+  /// A map storing access info of all structs
+  std::unordered_map<const Type*, StructFieldAccessInfo*> StructFieldAccessInfoMap;
+
+  /// \name Data structure to get statistics of each DebugStats entry
+  /// %{
+  std::vector<unsigned> StatCounts;
+
+  const std::vector<std::string> StatNames = {
+    "Variable type is Struct**",
+    "Function argument is a value",
+    "Function argument is not defined in the program",
+    "GEP value passed into function calls",
+    "GEP value passed into bitcast",
+    "GEP value passed into unexpected opcode",
+    "User is not Instruction nor Operator",
+    "Struct defined but no accesses",
+    "Struct passed into functions defined out of scope",
+    "GEP instruction directly used on struct*",
+    "Unknown instruction directly used on struct*"
+  };
+  /// %}
+
+  /// Used to print name of each StructDefinitionType
+  const std::vector<std::string> StructDefinitionTypeNames = {
+    "global struct",
+    "global struct*",
+    "global struct**",
+    "local struct",
+    "local struct*",
+    "local struct**"
+  };
+
+};
+
+/// This class is used to store all access information for each struct
+/// declared in the program. It records all loads and stores to all fields
+/// of the struct to provide essential information for cache-aware struct
+/// field analysis.
+class StructFieldAccessInfo
+{
+ public:
+  StructFieldAccessInfo(const Type* T, const StructFieldAccessManager::StructDefinitionType ST, const Module& MD, const StructFieldAccessManager* M, const DICompositeType* D):
+      Eligiblity(true),
+      CurrentModule(MD),
+      StructureType(NULL),
+      StructDefinition(ST),
+      DebugInfo(D),
+      NumElements(0),
+      StructManager(M),
+      StatCounts(StructFieldAccessManager::DebugStats::DS_MaxNumStats)
+  {
+    assert(T && T->isStructTy());
+    StructureType = dyn_cast<StructType>(T);
```

Comment:
Use cast instead of dyn_cast if you know it is always passed the right type. Actually, it is then preferable to cast into StructType at the caller of this constructor and make this take StructType *ST.